app/assets/javascripts/admin/addon/components/emoji-value-list.js

```diff
@@ -0,0 +1,186 @@
+import Component from "@ember/component";
+import I18n from "I18n";
+import { isEmpty } from "@ember/utils";
+import { on } from "discourse-common/utils/decorators";
+import { emojiUrlFor } from "discourse/lib/text";
+import { action, set } from "@ember/object";
+import { later, schedule } from "@ember/runloop";
+
+export default Component.extend({
+  classNameBindings: [":value-list"],
+  collection: null,
+  values: null,
+  validationMessage: null,
+  emojiPickerIsActive: false,
+  isEditorFocused: false,
+  emojiName: null,
+
+  init() {
+    this._super(...arguments);
+    this.set("collection", []);
+  },
+
+  @action
+  emojiSelected(code) {
+    this.set("emojiName", code);
+    this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    this.set("isEditorFocused", !this.isEditorFocused);
+  },
+
+  @action
+  openEmojiPicker() {
+    this.set("isEditorFocused", !this.isEditorFocused);
+    later(() => {
+      this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    }, 100);
+  },
+
+  @action
+  clearInput() {
+    this.set("emojiName", "");
+  },
+
+  @on("didReceiveAttrs")
+  _setupCollection() {
+    this.set("collection", this._splitValues(this.values));
+  },
+
+  _splitValues(values) {
+    if (values && values.length) {
+      const res = [];
+      const emojis = values.split("|");
+      emojis.forEach((emoji, index) => {
+        const object = {};
+        object.value = emoji;
+
+        object.emojiUrl = emojiUrlFor(emoji);
+        object.isEditable = true;
+
+        object.isEditing = false;
+
+        object.isLast = emojis.length - 1 === index;
+
+        res.push(object);
+      });
+
+      return res;
+    } else {
+      return [];
+    }
+  },
+
+  @action
+  editValue(index) {
+    const item = this.collection[index];
+    if (item.isEditable) {
+      set(item, "isEditing", !item.isEditing);
+      schedule("afterRender", () => {
+        const textbox = document.querySelector(
+          `[data-index="${index}"] .value-input`
+        );
+        if (textbox) {
+          textbox.focus();
+        }
+      });
+    }
+  },
+
+  @action
+  changeValue(index, newValue) {
+    const item = this.collection[index];
+
+    if (this._checkInvalidInput(newValue)) {
+      const oldValues = this.values.split("|");
+
+      set(item, "value", oldValues[index - 1]);
+      set(item, "isEditing", !item.isEditing);
+
+      return;
+    }
+
+    this._replaceValue(index, newValue);
+
+    set(item, "isEditing", !item.isEditing);
+  },
+
+  @action
+  addValue() {
+    if (this._checkInvalidInput([this.emojiName])) {
+      return;
+    }
+    this._addValue(this.emojiName);
+    this.set("emojiName", "");
+  },
+
+  @action
+  removeValue(value) {
+    this._removeValue(value);
+  },
+
+  @action
+  shiftUp(index) {
+    if (!index) {
```

Comment:
did you lose my comment about sharing a shift fn here ?

Reply:
I read the comment, but the reason for not doing this was, we can make it a single function but we are doing totally different operations in both of these functions. combining both will reduce the readability of the function IMO. thoughts?

```
shiftDirection(index, direction) {
    if (direction === "up" && !index) {
      return;
    }

    if (direction === "down" && index === this.collection.length) {
      return;
    }

    const temp = this.collection[index];

    if(direction === "up") {
      this.collection[index] = this.collection[index - 1];
      this.collection[index - 1] = temp;
    } else {
      this.collection[index] = this.collection[index + 1];
      this.collection[index + 1] = temp;
    }

    this._saveValues();
  }
```

Reply:
why not just passing the operator? +1 or - 1

Reply:
Sorry, I didn't got you

Reply:
```
shift(index, operation) {
  if (index === this.collection.length) {
    return;
  }
  const temp = this.collection[index];
  this.collection[index] = this.collection[index + operation];
  this.collection[index + operation] = temp;
  this._saveValues();
}

shiftDown(index) {
  this.shift(index, 1);
}

shiftUp(index) {
    this.shift(index, -1);
}
```

Reply:
How about this one? in the latest commit?

Reply:
why not having this logic in shift itself?

Reply:
There's some issue in circling with shifting, as we are shifting elements and not rotating when we circle elements it swaps first and last elements which is kinda weird behavior. so I think we should skip the circling effect. Users can reach any position from anywhere with the current effect too.


https://user-images.githubusercontent.com/35112518/111456678-4a0c3180-873d-11eb-964d-693e3594a371.mov

Reply:
@jjaffeux thoughts on this?

Reply:
well it's probably than your logic is wrong, thumbsdown should go second place here, not last

Reply:
@jjaffeux Sorry, I missed this comment but added rotation of list in the latest commit.


https://user-images.githubusercontent.com/35112518/112463776-acda7a00-8d88-11eb-949b-f45f8d88b5f4.mov